algosdk/atomic_transaction_composer.py

```diff
@@ -0,0 +1,529 @@
+    ):
+        """
+        Add a smart contract method call to this atomic group.
+
+        An error will be thrown if the composer's status is not BUILDING,
+        if adding this transaction causes the current group to exceed
+        MAX_GROUP_SIZE, or if the provided arguments are invalid for
+        the given method.
+
+        Args:
+            app_id (int): application id of app that the method is being invoked on
+            method (Method): ABI method object with initialized arguments and return types
+            sender (str): address of the sender
+            sp (SuggestedParams): suggested params from algod
+            signer (TransactionSigner): signer that will sign the transactions
+            method_args (list[ABIValue | TransactionWithSigner], optional): list of arguments to be encoded
+                or transactions that immediate precede this method call
+            on_complete (OnComplete, optional): intEnum representing what app should do on completion
+                and if blank, it will default to a NoOp call
+            note (bytes, optional): arbitrary optional bytes
+            lease (byte[32], optional): specifies a lease, and no other transaction
+                with the same sender and lease can be confirmed in this
+                transaction's valid rounds
+            rekey_to (str, optional): additionally rekey the sender to this address
+
+        """
+        if self.status != AtomicTransactionComposerStatus.BUILDING:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposer must be in BUILDING state for a transaction to be added"
+            )
+        if len(self.txn_list) + method.get_txn_calls() > self.MAX_GROUP_SIZE:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposer cannot exceed MAX_GROUP_SIZE transactions"
+            )
+        if not method_args:
+            method_args = []
+        if len(method.args) != len(method_args):
+            raise error.AtomicTransactionComposerError(
+                "number of method arguments do not match the method signature"
+            )
+        if not isinstance(method, abi.method.Method):
+            raise error.AtomicTransactionComposerError(
+                "invalid Method object was passed into AtomicTransactionComposer"
+            )
+        if app_id == 0:
+            raise error.AtomicTransactionComposerError(
+                "application create call not supported"
+            )
+
+        app_args = []
+        # For more than 14 args, including the selector, compact them into a tuple
+        additional_args = []
+        additional_types = []
+        txn_list = []
+        # First app arg must be the selector of the method
+        app_args.append(method.get_selector())
+        # Iterate through the method arguments and either pack a transaction
+        # or encode a ABI value.
+        for i, arg in enumerate(method.args):
```

Comment:
```suggestion
        for arg in method.args:
```
don't really need `i` (but you might actually want to use it, in which case just ignore this suggestion)

Reply:
I just need it for `method_args[i]`; `arg` is for `method.args`. I realize it's a bit visually confusing (originally, `method` was called `method_call` for some differentiation, but the name was misleading for a `Method` object).

Reply:
👍